OpenRA.Mods.Common/Traits/Player/ProvidesPrerequisite.cs

```diff
@@ -62,21 +58,44 @@ public IEnumerable<string> ProvidesPrerequisites
 			}
 		}
 
+		protected override void Created(Actor self)
+		{
+			techTree = self.Owner.PlayerActor.Trait<TechTree>();
+
+			base.Created(self);
+		}
+
 		public void OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
 		{
-			if (info.ResetOnOwnerChange)
+			techTree = newOwner.PlayerActor.Trait<TechTree>();
+
+			if (Info.ResetOnOwnerChange)
 				Update(newOwner, newOwner.Faction.InternalName);
 		}
 
 		void Update(Player owner, string faction)
 		{
-			enabled = true;
+			enabled = !IsTraitDisabled;
+			if (IsTraitDisabled)
+				return;
 
-			if (info.Factions.Any())
-				enabled = info.Factions.Contains(faction);
+			if (Info.Factions.Any())
+				enabled = Info.Factions.Contains(faction);
 
-			if (info.RequiresPrerequisites.Any() && enabled)
-				enabled = owner.PlayerActor.Trait<TechTree>().HasPrerequisites(info.RequiresPrerequisites);
+			if (Info.RequiresPrerequisites.Any() && enabled)
+				enabled = techTree.HasPrerequisites(Info.RequiresPrerequisites);
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			Update(self.Owner, self.Owner.Faction.InternalName);
```

Comment:
This will ignore the value of `ResetOnOwnerChange` and unconditionally reset the faction to the current owner when the trait is disabled/reenabled.  You will need to cache the faction in the ctor and use that instead if `ResetOnOwnerChange` is false.